crates/ruff_python_formatter/src/expression/expr_joined_str.rs

```diff
@@ -1,21 +1,43 @@
 use crate::context::PyFormatContext;
-use crate::expression::parentheses::{NeedsParentheses, OptionalParentheses};
-use crate::{not_yet_implemented_custom_text, FormatNodeRule, PyFormatter};
-use ruff_formatter::{write, Buffer, FormatResult};
+use crate::expression::parentheses::{NeedsParentheses, OptionalParentheses, Parentheses};
+use crate::expression::string::{StringPrefix, StringQuotes};
+use crate::expression::Expr;
+use crate::prelude::*;
+use crate::{verbatim_text, FormatNodeRule, PyFormatter};
+use ruff_formatter::{write, Buffer, FormatError, FormatResult};
 use ruff_python_ast::node::AnyNodeRef;
 use rustpython_parser::ast::ExprJoinedStr;
 
 #[derive(Default)]
 pub struct FormatExprJoinedStr;
 
 impl FormatNodeRule<ExprJoinedStr> for FormatExprJoinedStr {
-    fn fmt_fields(&self, _item: &ExprJoinedStr, f: &mut PyFormatter) -> FormatResult<()> {
-        write!(
-            f,
-            [not_yet_implemented_custom_text(
-                r#"f"NOT_YET_IMPLEMENTED_ExprJoinedStr""#
-            )]
-        )
+    fn fmt_fields(&self, item: &ExprJoinedStr, f: &mut PyFormatter) -> FormatResult<()> {
+        let ExprJoinedStr { range, values } = item;
+
+        // TODO: can we re-use instead of copy/paste?
+        let string_content = f.context().locator().slice(*range);
+        let prefix = StringPrefix::parse(string_content);
+        let after_prefix = &string_content[usize::from(prefix.text_len())..];
+
+        let quotes = StringQuotes::parse(after_prefix).ok_or(FormatError::syntax_error(
+            "Didn't find string quotes after prefix",
+        ))?;
+
+        prefix.fmt(f)?;
+        quotes.fmt(f)?;
+        // TODO: join_with empty? or some other join?
```

Comment:
You can use `f.join()` to join elements without a separator (assuming that's what we need). We should look into how black splits long fstrings (best to include examples where the f-string is used directly in an ExpressionStatement and once where it is inside of parentheses).

Does python allow comments inside of string expressions or is this only a JavaScript madness?

Reply:
```
>>> f'asd{#foo}'
  File "<stdin>", line 1
    f'asd{#foo}'
                ^
SyntaxError: f-string expression part cannot include '#'
```

at least not i py3.11

Reply:
atm i'm (or maybe this isn't the best approach?) focused on getting the test suite to run, ie on producing output that's 1) valid python and 2) stable

(thanks for `f.join()`)